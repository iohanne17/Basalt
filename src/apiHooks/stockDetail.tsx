import {
  useGetStockListDetailQuery,
  useLazyFetchDetailUsingRangeQuery,
} from '@Features/stockDetail/stockDetail-api-slice';
import {useState} from 'react';

interface IFindStockDetail {
  symbol: string;
  date_from: string;
  date_to: string;
}

export const useStockListDetail = (symbol: string) => {
  const [showLazyData, setShowLazyData] = useState(false);
  const {isLoading, isFetching, isError, data, refetch} =
    useGetStockListDetailQuery(symbol);
  const {findTrigger, apiResponse} = useLazyStockListDetail();

  const findStockDetail = async (val: IFindStockDetail) => {
    try {
      await findTrigger(val);
      if (apiResponse.isSuccess) {
        setShowLazyData(true);
      }
    } catch (error) {
      setShowLazyData(false);
      throw new Error('Failed to fetch');
    }
  };

  const refetchData = () => {
    setShowLazyData(false);
    refetch();
  };

  return {
    isFetching: isFetching,
    isError,
    isLoading: isLoading,
    data: data?.data,
    refetch: refetchData,
    findStockDetail,
    apiLazyResponse: apiResponse,
    lazyHandle: showLazyData,
  };
};

export const useLazyStockListDetail = () => {
  const [findTrigger, apiResponse] = useLazyFetchDetailUsingRangeQuery();

  return {
    isFetching: apiResponse.isFetching,
    data: apiResponse.data,
    findTrigger,
    apiResponse,
  };
};
