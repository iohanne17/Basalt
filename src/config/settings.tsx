/**
 * settings.js
 * Internal settings and configurations for various aspects of the mobile app project
 */

const namespace = 'basalt';

const devSettings = {
  namespace,
  carApi: 'http://api.carsxe.com',
  carApiKey: 'z06wfb44x_gwme38v13_dr7023njo',
  favouriteId: `${namespace}favourite-car`,
};

const prodSettings = {
  ...devSettings,
};

const settings =
  process.env.NODE_ENV === 'production' ? prodSettings : devSettings;
export default settings;
