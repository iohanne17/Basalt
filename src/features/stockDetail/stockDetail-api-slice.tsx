import {createApi, fetchBaseQuery} from '@reduxjs/toolkit/query/react';
import Settings from '@Config/settings';

interface objInput {
  symbol: string;
  date_from: string;
  date_to: string;
}

export const stockListDetailSliceApi = createApi({
  reducerPath: 'stock-detail-api',
  baseQuery: fetchBaseQuery({
    baseUrl: Settings.marketApi,
    prepareHeaders(headers) {
      return headers;
    },
  }),
  endpoints: build => ({
    getStockListDetail: build.query<void, unknown>({
      query: (symbol: string) =>
        `/tickers/${symbol}?access_key=${Settings.ApiKey}`,
    }),
    fetchDetailUsingRange: build.query<void, unknown>({
      query: (obj: objInput) =>
        `/eod/${obj.symbol}?access_key=${Settings.ApiKey}&date_from=${obj.date_from}&date_to=${obj.date_to}`,
    }),
  }),
});

export const {useGetStockListDetailQuery} = stockListDetailSliceApi;
