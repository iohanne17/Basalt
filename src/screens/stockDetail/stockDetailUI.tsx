import React from 'react';
import {ScrollView, StyleSheet, View} from 'react-native';
import {
  ErrorScreen,
  HeaderLayout,
  Loading,
  Row,
  Spacer,
  Text,
  TextColor,
} from '@Components/index';
import {NativeStackScreenProps} from '@react-navigation/native-stack';
import {DetailRoutes} from 'src/navigation/routes';
import {DetailRoutesParams} from 'src/navigation/types';
import useStockListDetail from '@Hooks/stockDetail';
import {Theme} from '@Theme/theme';

export type DetailScreenProps = NativeStackScreenProps<
  DetailRoutesParams,
  DetailRoutes.STOCKDETAIL
>;

export interface StockInfo {
  adj_close: number;
  adj_high: number;
  adj_low: number;
  adj_open: number;
  adj_volume: number;
  close: number;
  date: Date;
  dividend: number;
  exchange: string;
  high: number;
  low: number;
  open: number;
  split_factor: number;
  symbol: string;
  volume: number;
}

export const StockDetail = ({route}: DetailScreenProps) => {
  const {symbol, icon, name} = route?.params;

  const {isLoading, isFetching, isError, data, refetch} =
    useStockListDetail(symbol);

  if (isLoading || isFetching) {
    return <Loading />;
  }

  if (isError) {
    return (
      <ErrorScreen
        title="Data Error"
        message="Error fetching data... Try again later"
        onPress={refetch}
        isLoading={isLoading}
      />
    );
  }

  const response = data ? data?.data[0] : undefined;
  const entries = response ? Object?.entries(response) : undefined;
  const detailArray = entries
    ? entries.slice(0, entries.length - 1)
    : undefined;

  return (
    <HeaderLayout
      headerTitle={symbol}
      headerShown={true}
      safeAreaStyle={s.safeAreaStyle}
      innerStyle={s.container}>
      <Spacer height={20} />
      <View style={s.roundIcon}>
        <Text title_5 color={TextColor.black}>
          {icon}
        </Text>
      </View>
      <View style={s.center}>
        <Text text style={s.text}>
          {name}
        </Text>
      </View>
      <Spacer height={20} />
      <ScrollView contentContainerStyle={s.detailContainer}>
        {!detailArray ? (
          <ErrorScreen
            title={'Unable to get data'}
            message="Try again"
            showButton={false}
          />
        ) : (
          detailArray.map((el, index) => {
            let [key, value] = el;
            return (
              <Row
                title={key}
                value={value as string | number}
                key={`${index}`}
              />
            );
          })
        )}
      </ScrollView>
    </HeaderLayout>
  );
};

export default StockDetail;

const s = StyleSheet.create({
  container: {
    flex: 1,
    backgroundColor: '#fff',
  },
  safeAreaStyle: {
    backgroundColor: Theme.colors.light.offwhite,
  },
  iconContainer: {
    alignItems: 'center',
    justifyContent: 'center',
    flexDirection: 'row',
    flex: 1,
  },
  roundIcon: {
    height: Theme.sizes.bigCircleHeight + 20,
    width: Theme.sizes.bigCircleWidth + 20,
    backgroundColor: Theme.colors.light.primary,
    alignItems: 'center',
    justifyContent: 'center',
    alignSelf: 'center',
  },
  text: {
    color: Theme.colors.dark.inverseWhite100,
    textTransform: 'capitalize',
  },
  center: {
    flexDirection: 'row',
    justifyContent: 'center',
    alignItems: 'center',
  },
  detailContainer: {
    ...StyleSheet.absoluteFillObject,
    borderTopLeftRadius: Theme.sizes.r4,
    borderTopRightRadius: Theme.sizes.r4,
    flex: 1,
    backgroundColor: Theme.colors.light.offwhite,
    paddingHorizontal: Theme.paddings.h16,
    paddingVertical: Theme.paddings.h16,
    top: 5,
  },
  errorText: {
    textTransform: 'uppercase',
    textAlign: 'center',
    marginTop: Theme.sizes.bigCircleHeight,
  },
});
