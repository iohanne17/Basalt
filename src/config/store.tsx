import {configureStore} from '@reduxjs/toolkit';
import {stockListSliceApi} from '@Features/stockList/stockList-api-slice';
import {stockListDetailSliceApi} from '@Features/stockDetail/stockDetail-api-slice';
import {TypedUseSelectorHook, useDispatch, useSelector} from 'react-redux';
import {setupListeners} from '@reduxjs/toolkit/dist/query';

export const store = configureStore({
  reducer: {
    [stockListSliceApi.reducerPath]: stockListSliceApi.reducer,
    [stockListDetailSliceApi.reducerPath]: stockListDetailSliceApi.reducer,
  },
  middleware: getDefaultMiddle => {
    return getDefaultMiddle().concat(
      stockListSliceApi.middleware,
      stockListDetailSliceApi.middleware,
    );
  },
});

export type AppDispatch = typeof store.dispatch;
export type RootState = ReturnType<typeof store.getState>;
export const useTypedDispatch = () => useDispatch<AppDispatch>();
export const useTypedSelector: TypedUseSelectorHook<RootState> = useSelector;

setupListeners(store.dispatch);
