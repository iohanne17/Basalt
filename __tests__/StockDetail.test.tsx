import React from 'react'
import { render } from '@testing-library/react-native'
import { StockDetail, DetailScreenProps } from '../src/screens/stockDetail/stockDetailUI'
import { Provider } from 'react-redux'
import { store } from '../src/config/store'

const useStockListDetail = jest.fn()

const sample = {
  adj_close: 900,
  adj_high: 90,
  adj_low: 0.8,
  adj_open: 877,
  adj_volume: 9000,
  close: 9888,
  dividend: 900,
  exchange: 90000,
  high: 208,
  low: 18,
  open: 870,
  split_factor: 678,
  symbol: 'AAPL',
  volume: 4,
}

// Mock the useStockListDetail hook
jest.mock('@Hooks/stockDetail', () => ({
  __esModule: true,
  default: jest.fn(),
  useStockListDetail: jest.fn(() => ({
    isLoading: false,
    isFetching: false,
    isError: false,
    data: {
      data: [sample, sample],
    },
    refetch: jest.fn(),
  })),
}))

jest.mock('@Components/index', () => ({
  ErrorScreen: 'ErrorScreen',
  HeaderLayout: 'HeaderLayout',
  Loading: 'Loading',
  Spacer: 'Spacer',
  Text: 'Text',
  TextColor: 'TextColor',
}))

describe('Detail Component', () => {
  let navigation
  beforeEach(() => {
    navigation = {
      navigation: {
        navigate: jest.fn(),
      },
      route: {
        params: {
          symbol: 'AAPL',
          icon: 'apple',
          name: 'Aplle Inc',
        },
      },
    }

    useStockListDetail.mockClear()
  })

  useStockListDetail.mockReturnValue({
    isLoading: false,
    isFetching: false,
    isError: false,
    data: {
      data: [sample, sample],
    },
    refetch: jest.fn(),
  })

  it('matches stock detail snapshot', () => {
    const { toJSON } = render(
      <Provider store={store}>
        <StockDetail {...(navigation as DetailScreenProps)} />
      </Provider>,
    )
    expect(toJSON()).toMatchSnapshot()
  })
})
