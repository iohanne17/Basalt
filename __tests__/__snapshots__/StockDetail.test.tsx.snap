// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Detail Component matches stock detail snapshot 1`] = `
<HeaderLayout
  headerShown={true}
  headerTitle="AAPL"
  innerStyle={
    Object {
      "backgroundColor": "#fff",
      "flex": 1,
    }
  }
  safeAreaStyle={
    Object {
      "backgroundColor": "#FAF9F6",
    }
  }
>
  <Spacer
    height={20}
  />
  <View
    style={
      Object {
        "alignItems": "center",
        "alignSelf": "center",
        "backgroundColor": "#FE812A",
        "height": 70,
        "justifyContent": "center",
        "width": 70,
      }
    }
  >
    <Text
      title_5={true}
    >
      apple
    </Text>
  </View>
  <View
    style={
      Object {
        "alignItems": "center",
        "flexDirection": "row",
        "justifyContent": "center",
      }
    }
  >
    <Text
      style={
        Object {
          "color": "#000000",
          "textTransform": "capitalize",
        }
      }
      text={true}
    >
      Aplle Inc
    </Text>
  </View>
  <Spacer
    height={20}
  />
  <RCTScrollView
    contentContainerStyle={
      Object {
        "backgroundColor": "#FAF9F6",
        "borderTopLeftRadius": 25,
        "borderTopRightRadius": 25,
        "bottom": 0,
        "flex": 1,
        "left": 0,
        "paddingHorizontal": 16,
        "paddingVertical": 16,
        "position": "absolute",
        "right": 0,
        "top": 5,
      }
    }
  >
    <View>
      <ErrorScreen
        message="Try again"
        showButton={false}
        title="Unable to get data"
      />
    </View>
  </RCTScrollView>
</HeaderLayout>
`;
